retrofit-adapters/rxjava/src/main/java/retrofit2/adapter/rxjava/RxJavaCallAdapterFactory.java

```diff
@@ -131,14 +133,28 @@ private RxJavaCallAdapterFactory(Scheduler scheduler) {
 
     @Override public void call(final Subscriber<? super Response<T>> subscriber) {
       // Since Call is a one-shot type, clone it for each new subscriber.
-      final Call<T> call = originalCall.clone();
+      Call<T> call = originalCall.clone();
 
-      // Attempt to cancel the call if it is still in-flight on unsubscription.
-      subscriber.add(Subscriptions.create(new Action0() {
-        @Override public void call() {
-          call.cancel();
-        }
-      }));
+      // Wrap the call in a helper which handles both unsubscription and backpressure.
+      RequestArbiter<T> requestArbiter = new RequestArbiter<>(call, subscriber);
+      subscriber.add(Subscriptions.create(requestArbiter));
+      subscriber.setProducer(requestArbiter);
+    }
+  }
+
+  static final class RequestArbiter<T> extends AtomicBoolean implements Action0, Producer {
```

Comment:
Saves space at runtime. It's a common practice in RxJava.

On Tue, Mar 29, 2016, 9:38 PM Jesse Wilson <notifications@github.com> wrote:

> In
> retrofit-adapters/rxjava/src/main/java/retrofit2/adapter/rxjava/RxJavaCallAdapterFactory.java
> <https://github.com/square/retrofit/pull/1702#discussion_r57827069>:
>
> >
> > -      // Attempt to cancel the call if it is still in-flight on unsubscription.
> > -      subscriber.add(Subscriptions.create(new Action0() {
> > -        @Override public void call() {
> > -          call.cancel();
> > -        }
> > -      }));
> > +      // Wrap the call in a helper which handles both unsubscription and backpressure.
> > +      RequestArbiter<T> requestArbiter = new RequestArbiter<>(call, subscriber);
> > +      subscriber.add(Subscriptions.create(requestArbiter));
> > +      subscriber.setProducer(requestArbiter);
> > +    }
> > +  }
> > +
> > +  static final class RequestArbiter<T> extends AtomicBoolean implements Action0, Producer {
>
> you can extend AtomicBoolean? weird!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/square/retrofit/pull/1702/files/c107778914815466bda00878c4cbc6b799a6e133#r57827069>
>